File: Data Structures & Algorithms/daily-temperatures/submission-0.js
class Solution {
    /**
     * @param {number[]} temperatures
     * @return {number[]}
     */
    dailyTemperatures(temperatures) {
        let stack = []
        let ans = new Array(temperatures.length).fill(0)

        for (let i = 0; i < temperatures.length; i++) {
            
            while(stack.length && 
            temperatures[stack[stack.length - 1]] < temperatures[i]) {
                let idx = stack.pop()
                ans[idx] = i - idx
            }
            
            stack.push(i)
        }

        return ans;

    }
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int n = heights.size();

        vector<int> nextSmaller(n, n);
        vector<int> prevSmaller(n, -1);

        stack<int> st;

        // Next Smaller Element
        for (int i = 0; i < n; i++) {
            while (!st.empty() && heights[st.top()] > heights[i]) {
                nextSmaller[st.top()] = i;
                st.pop();
            }
            st.push(i);
        }

        while (!st.empty()) st.pop();

        // Previous Smaller Element
        for (int i = n - 1; i >= 0; i--) {
            while (!st.empty() && heights[st.top()] > heights[i]) {
                prevSmaller[st.top()] = i;
                st.pop();
            }
            st.push(i);
        }

        int ans = 0;

        for (int i = 0; i < n; i++) {
            int width = nextSmaller[i] - prevSmaller[i] - 1;
            ans = max(ans, heights[i] * width);
        }

        return ans;
    }
};